libs/cuda_support/include/hpx/cuda_support/cuda_executor.hpp

```diff
@@ -0,0 +1,206 @@
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/basic_execution/execution.hpp>
+#include <hpx/cuda_support/cuda_exception.hpp>
+#include <hpx/cuda_support/cuda_future.hpp>
+#include <hpx/cuda_support/target.hpp>
+#include <hpx/errors/exception.hpp>
+#include <hpx/execution/traits/is_executor.hpp>
+#include <hpx/futures/future.hpp>
+#include <hpx/futures/traits/future_access.hpp>
+
+// CUDA runtime
+#include <cuda_runtime.h>
+//
+#include <sstream>
+
+namespace hpx { namespace cuda {
+
+    namespace detail {
+        // -------------------------------------------------------------------------
+        // A helper object to call a cudafunction returning a cudaError type
+        // or a plain kernel definition (or cublas function in cublas executor)
+        template <typename R, typename... Args>
+        struct dispatch_helper;
+
+        // default implementation - call the function
+        template <typename R, typename... Args>
+        struct dispatch_helper
+        {
+            inline R operator()(R (*f)(Args...), Args... args)
+            {
+                return f(args...);
+            }
+        };
+
+        // specialization for return type void
+        template <typename... Args>
+        struct dispatch_helper<void, Args...>
+        {
+            inline void operator()(void (*f)(Args...), Args... args)
+            {
+                f(args...);
+            }
+        };
+
+        // specialization for return type of cudaError_t
+        template <typename... Args>
+        struct dispatch_helper<cudaError_t, Args...>
+        {
+            inline cudaError_t operator()(
+                cudaError_t (*f)(Args...), Args... args)
+            {
+                cudaError_t err = f(args...);
+                return check_cuda_error(err);
+            }
+        };
+
+    }    // namespace detail
+
+    // -------------------------------------------------------------------------
+    // Allows the launching of cuda functions and kernels on a stream with futures
+    // returned that are set when the async functions/kernels are ready
+    // -------------------------------------------------------------------------
+    struct cuda_executor_base
+    {
+        using future_type = hpx::future<void>;
+
+        cuda_executor_base(cuda_executor_base& other) = delete;
+        cuda_executor_base(const cuda_executor_base& other) = delete;
+        cuda_executor_base operator=(const cuda_executor_base& other) = delete;
+
+        // -------------------------------------------------------------------------
+        // constructors - create a cuda stream that all tasks invoked by
+        // this helper will use
+        // assume event mode is the default
+        cuda_executor_base(std::size_t device, bool event_mode)
+          : device_(device)
+          , event_mode_(event_mode)
+          , target_(device)
```

Comment:
`target_` seems to be an abstraction over a `cudaStream_t`. This is similar to my question for the `cublas_executor`, there is some overhead with creating and destroying cudaStream_t as discussed at PyTorch : https://github.com/pytorch/pytorch/issues/9646 . Have you considered making a pool of cudaStream_t which the executor can get a stream from instead?

Reply:
Same response really. In octotiger, a collection of (say) 128 streams are created during an initialization step and then when needed, they are handed out in sequentially so that work can be pushed into the GPU and it can be oversubscribed - when all streams are 'busy' - further work can be blocked or the cpu can be given tasks. The streams can be on a single GPU, or divided over multiple ones and synchronization between streams can either use futures or cuda stream semantics - depending on the workflow.

Something like this can be implemented using the targets, or executors as handles onto a preallocated set of streams/handles. `targets` are a hold over from `hpx::compute` and they don't really buy us anything useful, but for this implementation I though it simpler to reuse them rather than getting rid of them and reimplementing a new abstraction over streams. The concept of a `target` is quite nice though. We can/should work on efficient management of streams/handles for the next PR.